monitoring/metric/counter.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+type Counter interface {
+	Add(n int64)
+}
+type counter struct {
+	mu              sync.Mutex
+	name            string
+	value           int64
+	lastDumped      time.Time
+	lastDumpedValue int64
+}
+
+type safeMetrics struct {
+	mu sync.Mutex
+	m  map[string]*counter
+}
+
+var (
+	metrics = safeMetrics{
```

Comment:
The model is that code everywhere can just create their metrics, without any knowledge of where they should go, just caring about updating them, and then, their values can be queried and published in different ways (the `DumpToLog` is the simplistic one I'm providing initially, but there would be an HTTP handler that would return them in a format suitable for Prometheus, for example).